packages/liferay-npm-scripts/src/scripts/storybook.js

```diff
@@ -0,0 +1,148 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+
+const getMergedConfig = require('../utils/getMergedConfig');
+const log = require('../utils/log');
+const spawnSync = require('../utils/spawnSync');
+
+const BABEL_CONFIG = getMergedConfig('babel');
+const STORYBOOK_CONFIG = getMergedConfig('npmscripts').storybook;
+
+const NODE_PATHS = [
+	path.join(__dirname, '../../../../node_modules'),
+	path.join(__dirname, '../../node_modules')
+];
+
+const STORYBOOK_CONFIG_DIR_PATH = path.join(__dirname, '../storybook');
+
+const STORYBOOK_CONFIG_FILES = [
+	'addons.js',
+	'config.js',
+	'middleware.js',
+	'preview-head.html',
+	'webpack.config.js'
+];
+
+const PORTAL_ROOT = process.cwd().split('/modules')[0];
+
+/**
+ * Builds the string used for defining the NODE_PATH environment.
+ * On Windows, NODE_PATH is delimited by semicolons (;) instead of colons.
+ * @returns {String}
+ */
+function buildNodePath() {
+	return NODE_PATHS.join(process.platform === 'win32' ? ';' : ':');
+}
+
+/**
+ * Copies storybook config files into the build path.
+ * @param {String} buildPath Path to the build directory.
+ * @param {Array} files The list of files to copy.
+ */
+function copyStorybookConfigFiles(buildPath, files) {
+	files.forEach(function(file) {
+		fs.copyFileSync(
+			path.join(STORYBOOK_CONFIG_DIR_PATH, file),
+			path.join(buildPath, file)
+		);
+	});
+}
+
+/**
+ * Combine multiple Language.properties files to a single file for
+ * liferay-lang-key-dev-loader to read from.
+ * @param {String} buildPath Path to the build directory.
+ * @param {Array} paths List of path strings to Language.properties files.
+ */
+function compileLanguageProperties(buildPath, paths) {
+	const PORTAL_LANG_PATH = path.join(
+		PORTAL_ROOT,
+		'/portal-impl/src/content/Language.properties'
+	);
+
+	const LANG_PATHS = [PORTAL_LANG_PATH, ...paths];
+
+	// Used for keeping track of the current file being processed to show the
+	// correct file that caused an error.
+	let currentLangPath;
+
+	try {
+		const bufferArray = LANG_PATHS.filter(path => fs.existsSync(path)).map(
+			path => {
+				currentLangPath = path;
+
+				return Buffer.concat([
+					fs.readFileSync(path),
+					Buffer.from('\n')
+				]);
+			}
+		);
+
+		const buffer = Buffer.concat(bufferArray);
+
+		fs.writeFileSync(
+			path.join(buildPath, 'Language.properties'),
+			buffer.toString('utf8')
+		);
+	} catch (e) {
+		log(`Failed to read lang key file: ${currentLangPath}`);
+
+		// Write an empty file to prevent liferay-lang-key-dev-loader from
+		// breaking the build if no file is found.
+		fs.writeFileSync(path.join(buildPath, 'Language.properties'), '');
+	}
```

Comment:
One idea here, you could avoid the `let currentLangPath` stuff by moving the `try`/`catch` inside the loop. Apart from wanting to avoid an unnecessary temporary variable, keeping the contents of a `try` as narrowly scoped as possible seems like a good idea as a general practice.

```javascript
let buffer = Buffer.alloc(0);

LANG_PATHS.filter(fs.existsSync).forEach(langPath => {
  try {
    buffer = Buffer.concat([
      buffer,
      fs.readFileSync(langPath),
      Buffer.from('\n')
    ]); 
  } catch (error)
    log(`Failed to read lang key file "${langPath}" due to error: ${error}`);
  }
});

fs.writeFileSync(
  path.join(buildPath, 'Language.properties'),
  buffer.toString('utf8')
);
```

As far as I can tell, that would mean you'd get (at least) your empty file no matter what, so you wouldn't even need the comment calling it out.

I also wonder though whether all this needs to be done using buffers explicitly. Would just strings suffice? eg.

```javascript
let output = [];

LANG_PATHS.filter(fs.existsSync).forEach(langPath => {
  try {
    output.push(fs.readFileSync(langPath).toString());
  } catch (error)
    log(`Failed to read lang key file "${langPath}" due to error: ${error}`);
  }
});

fs.writeFileSync(
  path.join(buildPath, 'Language.properties'),
  output.join('\n')
);
```

Reply:
Tested out with strings instead and it works. 👌 Looks much cleaner so I'll go with that approach. I'll keep in mind the structuring of the try/catch blocks though!